cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/prometheus/PrometheusMetricSampler.java

```diff
@@ -0,0 +1,269 @@
+            try {
+                _samplingIntervalMs = Integer.parseInt(samplingIntervalMsString);
+            } catch (NumberFormatException e) {
+                throw new ConfigException("%s config should be a positive number, provided %s",
+                    samplingIntervalMsString);
+            }
+
+            if (_samplingIntervalMs <= 0) {
+                throw new ConfigException(String.format("%s config should be set to positive,"
+                                                        + " provided %d.",
+                                                        PROMETHEUS_QUERY_RESOLUTION_STEP_MS_CONFIG,
+                                                        _samplingIntervalMs));
+            }
+        }
+    }
+
+    private void configurePrometheusAdapter(Map<String, ?> configs) {
+        final String endpoint = (String) configs.get(PROMETHEUS_SERVER_ENDPOINT_CONFIG);
+        if (endpoint == null) {
+            throw new ConfigException(String.format(
+                "%s config is required by Prometheus metric sampler", PROMETHEUS_SERVER_ENDPOINT_CONFIG));
+        }
+
+        try {
+            HttpHost host = HttpHost.create(endpoint);
+            if (host.getPort() < 0) {
+                throw new IllegalArgumentException();
+            }
+            CloseableHttpClient httpClient = HttpClients.createDefault();
+            _prometheusAdapter = new PrometheusAdapter(httpClient, host, _samplingIntervalMs);
+        } catch (IllegalArgumentException ex) {
+            throw new ConfigException(
+                String.format("Prometheus endpoint URI is malformed, "
+                              + "expected schema://host:port, provided %s", endpoint));
+        }
+    }
+
+    private void configureQueryMap(Map<String, ?> configs) {
+        String prometheusQuerySupplierClassName = (String) configs.get(PROMETHEUS_QUERY_SUPPLIER_CONFIG);
+        Class<?> prometheusQuerySupplierClass = DEFAULT_PROMETHEUS_QUERY_SUPPLIER;
+        if (prometheusQuerySupplierClassName != null) {
+            prometheusQuerySupplierClass = (Class<?>) ConfigDef.parseType(PROMETHEUS_QUERY_SUPPLIER_CONFIG,
+                prometheusQuerySupplierClassName, CLASS);
+            if (!PrometheusQuerySupplier.class.isAssignableFrom(prometheusQuerySupplierClass)) {
+                throw new ConfigException(String.format(
+                    "Invalid %s is provided to prometheus metric sampler, provided %s",
+                    PROMETHEUS_QUERY_SUPPLIER_CONFIG, prometheusQuerySupplierClass));
+            }
+        }
+        PrometheusQuerySupplier prometheusQuerySupplier = KafkaCruiseControlConfigUtils.getConfiguredInstance(
+            prometheusQuerySupplierClass, PrometheusQuerySupplier.class, Collections.emptyMap());
+        _metricToPrometheusQueryMap = prometheusQuerySupplier.get();
+    }
+
+    @Override
+    public void close() {
+        // do nothing
```

Comment:
I decided to make the `httpClient` a field, and close it here.

Is there any advantage to clearing the `Map`s here? They are going to be garbage-cleaned anyway.
I guess it makes sense to close open handles, etc. which is why I closed the `httpClient`.

Reply:
> Is there any advantage to clearing the Maps here? They are going to be garbage-cleaned anyway.

@wyuka TL;DR. I think there is no need to apply the suggestion I made.
-- more -- 
Garbage collection (GC) would trigger only when JVM thinks it needs a GC based on the heap size. Also, GC will not trigger as long as there are references to `PrometheusMetricSampler`; hence, it can potentially keep unused resources longer. That being said, clearing resources in `close()` will also not trigger GC, but will make those resources available for GC _earlier_. However, presumably references to `PrometheusMetricSampler` are not expected to linger around for a long time after it is closed. Thus, please feel free to ignore my suggestion.